### DISCUSSION TITLE
Is there a way to list all  the substructure of the mol?

### DISCUSSION THREAD

**Discussion by yangjinhaoo** [rdkit/rdkit#5382] [General]
rdkit postgresql plugin is great , but in the billions database, use UserDefineType to search substructure is very slow. I even need one day to execute a query. So I want to make it faster. Use substructure list and  "IN" is a way to  instead of "hassubstructure". But I can't find the function in rdkit. 
So is there a way to list all the substructure?

**Comment by greglandrum** [comment]
@yangjinhaoo What do you mean by "list all the substructures"? Do you want to do multiple substructure searches across a database?

**Reply by yangjinhaoo** [reply]
> @yangjinhaoo What do you mean by "list all the substructures"? Do you want to do multiple substructure searches across a database?

no, I want to search substructure in a billions database. I don't know how rdkit to work. But I think there is a func  and it need to real-time compute. I think it's unnecessary and it's a reason that makes my substructure search is so slow. So if I can list all the substructure ,  real-time compute time can be saved. And I also think substructure is a way to set inverted index.

**Comment by iwatobipen** [comment]
Hi @yangjinhaoo,

I think it worth to check following blog post.
http://rdkit.blogspot.com/2018/02/introducing-substructlibrary.html
rdSubstructLibrary works very fast.
Best,

Taka

**Reply by yangjinhaoo** [reply]
> Hi @yangjinhaoo,
> 
> I think it worth to check following blog post. http://rdkit.blogspot.com/2018/02/introducing-substructlibrary.html rdSubstructLibrary works very fast. Best,
> 
> Taka

it's awesome!  
It seem like provide a little memory database to search the substructure. But I don't think is useful for me .  I have half an  billion smiles.
I think I can't set them in memory all.  And I don't only use the substructure as select params. I also use some filter to pass the smiles, like mol_weight and so on .   And I will use similarity search too. So an extra memory database is not I want.
I think what I real need is find another way to add the function on the rdkit-postgresql to make substructure search faster. 
List all the substructure and use "IN"  operator is the ony way, I think. And I think it is easy, just a breadth traversal, maybe I think things are too simple..
But thank you.

**Comment by DavidACosgrove** [comment]
Generating all substructures in a database of any size is really not
feasible. If you can’t store all the SMILES in memory, you certainly won’t
be able to store all the sub-SMILES.  Try googling “Sayle small world” for
a discussion of how to do this and why it is unlikely to work for you.

On Tue, 28 Jun 2022 at 15:07, yangjinhaoo ***@***.***> wrote:

> Hi @yangjinhaoo <https://github.com/yangjinhaoo>,
>
> I think it worth to check following blog post.
> http://rdkit.blogspot.com/2018/02/introducing-substructlibrary.html
> rdSubstructLibrary works very fast. Best,
>
> Taka
>
> it's awesome!
> It seem like provide a little memory database to search the substructure.
> But I don't think is useful for me . I have half an billion smiles.
> I think I can't set them in memory all. And I don't only use the
> substructure as select params. I also use some filter to pass the smiles,
> like mol_weight and so on . And I will use similarity search too. So an
> extra memory database is not I want.
> I think what I real need is find another way to add the function on the
> rdkit-postgresql to make substructure search faster.
> List all the substructure and use "IN" operator is the ony way, I think.
> And I think it is easy, just a breadth traversal, maybe I think things are
> too simple..
> But thank you.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/5382#discussioncomment-3040021>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FVCKBAMKGIYH4FGOXTVRL2JNANCNFSM5ZOSA7QA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by yangjinhaoo** [reply]
> Generating all substructures in a database of any size is really not feasible. If you can’t store all the SMILES in memory, you certainly won’t be able to store all the sub-SMILES. Try googling “Sayle small world” for a discussion of how to do this and why it is unlikely to work for you.
> On Tue, 28 Jun 2022 at 15:07, yangjinhaoo ***@***.***> wrote: Hi @yangjinhaoo <https://github.com/yangjinhaoo>, I think it worth to check following blog post. http://rdkit.blogspot.com/2018/02/introducing-substructlibrary.html rdSubstructLibrary works very fast. Best, Taka it's awesome! It seem like provide a little memory database to search the substructure. But I don't think is useful for me . I have half an billion smiles. I think I can't set them in memory all. And I don't only use the substructure as select params. I also use some filter to pass the smiles, like mol_weight and so on . And I will use similarity search too. So an extra memory database is not I want. I think what I real need is find another way to add the function on the rdkit-postgresql to make substructure search faster. List all the substructure and use "IN" operator is the ony way, I think. And I think it is easy, just a breadth traversal, maybe I think things are too simple.. But thank you. — Reply to this email directly, view it on GitHub <[#5382 (reply in thread)](https://github.com/rdkit/rdkit/discussions/5382#discussioncomment-3040021)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACGF2FVCKBAMKGIYH4FGOXTVRL2JNANCNFSM5ZOSA7QA> . You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>
> -- David Cosgrove Freelance computational chemistry and chemoinformatics developer http://cozchemix.co.uk

I think I konw what you mean. But substructureLibrary store the smiles. If I can generate all the substructure, I just need store the List[string], I think it need less space, and It can be well integrated with the postgresql.

**Comment by yangjinhaoo** [comment]
I think I konw what you mean. But substructureLibrary store the smiles. If I can generate all the substructure, I just need store the List[string], I think it need less space, and It can be well integrated with the postgresql.
________________________________
发件人: David Cosgrove ***@***.***>
发送时间: 2022年6月28日 21:37
收件人: rdkit/rdkit ***@***.***>
抄送: Yang Jinhao ***@***.***>; Mention ***@***.***>
主题: Re: [rdkit/rdkit] Is there a way to list all the substructure of the mol? (Discussion #5382)


Generating all substructures in a database of any size is really not
feasible. If you can’t store all the SMILES in memory, you certainly won’t
be able to store all the sub-SMILES. Try googling “Sayle small world” for
a discussion of how to do this and why it is unlikely to work for you.

On Tue, 28 Jun 2022 at 15:07, yangjinhaoo ***@***.***> wrote:

> Hi @yangjinhaoo <https://github.com/yangjinhaoo>,
>
> I think it worth to check following blog post.
> http://rdkit.blogspot.com/2018/02/introducing-substructlibrary.html
> rdSubstructLibrary works very fast. Best,
>
> Taka
>
> it's awesome!
> It seem like provide a little memory database to search the substructure.
> But I don't think is useful for me . I have half an billion smiles.
> I think I can't set them in memory all. And I don't only use the
> substructure as select params. I also use some filter to pass the smiles,
> like mol_weight and so on . And I will use similarity search too. So an
> extra memory database is not I want.
> I think what I real need is find another way to add the function on the
> rdkit-postgresql to make substructure search faster.
> List all the substructure and use "IN" operator is the ony way, I think.
> And I think it is easy, just a breadth traversal, maybe I think things are
> too simple..
> But thank you.
>
> ―
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/5382#discussioncomment-3040021>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FVCKBAMKGIYH4FGOXTVRL2JNANCNFSM5ZOSA7QA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
--
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

―
